Sources/SourceControl/GitRepository.swift

```diff
@@ -325,9 +325,9 @@ public class GitRepository: Repository, WorkingCheckout {
 
     public func hasUnpushedCommits() throws -> Bool {
         return try queue.sync {
-            let hasOutput = try Process.checkNonZeroExit(
-                args: Git.tool, "-C", path.asString, "log", "--branches", "--not", "--remotes").chomp().isEmpty
-            return !hasOutput
+            let output = try Process.checkNonZeroExit(
+                args: Git.tool, "-C", path.asString, "branch", "-r", "--contains", "HEAD").chomp()
```

Comment:
The current logic check if there are any branches which are not on remote. So that covers most of the cases. If we can't find an easy way to detect the commits that are not on a branch, we should probably not worry about that case a lot (because it is a highly corner case).